File: e289(beautiful_ribbon)/Python/beautiful_ribbon.py
from sys import stdin
while True:
    line = stdin.readline()
    if not line: break
    m, n = map(int, line.split())
    ribbon = list(map(int, stdin.readline().split()))
    beauty = 0
    sub = {}
    for i in range(m):
        if ribbon[i] not in sub:
            sub[ribbon[i]] = 1
        else: sub[ribbon[i]] += 1
    if len(sub) == m: beauty += 1
    for i in range(m, n):
        if sub[ribbon[i - m]] > 1: sub[ribbon[i - m]] -= 1
        else: del sub[ribbon[i - m]]
        if ribbon[i] not in sub:
            sub[ribbon[i]] = 1
        else: sub[ribbon[i]] += 1
        if len(sub) == m: beauty += 1
    print(beauty)

File: e289(beautiful_ribbon)/C++/beautiful_ribbon.cpp
#include <iostream>
#include <map>

using namespace std;

int main(int argc, char const *argv[]) {
  ios_base::sync_with_stdio(false);
  cin.tie(0);

  int m, n;
  while (cin >> m >> n) {
    map<string, int> sub;
    string ribbon[n];
    string color;
    int beauty = 0;
    for (int i = 0; i < m; i++) {
      cin >> ribbon[i];
      sub[ribbon[i]] += 1;
    }
    if (sub.size() == m) beauty++;
    for (int i = m; i < n; i++) {
      if (sub[ribbon[i - m]] > 1) sub[ribbon[i - m]] -= 1;
      else sub.erase(ribbon[i - m]);
      cin >> ribbon[i];
      sub[ribbon[i]] += 1;
      if (sub.size() == m) beauty++;
    }
    cout << beauty << '\n';
  }
  return 0;
}
